File: app/src/main/java/com/kelsos/mbrc/ui/activities/nav/MainActivity.java
package com.kelsos.mbrc.ui.activities.nav;

import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.Typeface;
import android.os.Bundle;
import android.support.annotation.ColorRes;
import android.support.annotation.DrawableRes;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.support.v4.view.MenuItemCompat;
import android.support.v7.widget.ShareActionProvider;
import android.text.TextUtils;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.SeekBar;
import android.widget.TextView;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.OnLongClick;
import com.kelsos.mbrc.R;
import com.kelsos.mbrc.annotations.Connection;
import com.kelsos.mbrc.annotations.PlayerState;
import com.kelsos.mbrc.annotations.PlayerState.State;
import com.kelsos.mbrc.annotations.Repeat;
import com.kelsos.mbrc.annotations.Repeat.Mode;
import com.kelsos.mbrc.constants.Const;
import com.kelsos.mbrc.constants.Protocol;
import com.kelsos.mbrc.constants.ProtocolEventType;
import com.kelsos.mbrc.constants.UserInputEventType;
import com.kelsos.mbrc.data.UserAction;
import com.kelsos.mbrc.domain.TrackInfo;
import com.kelsos.mbrc.enums.LfmStatus;
import com.kelsos.mbrc.events.MessageEvent;
import com.kelsos.mbrc.events.bus.RxBus;
import com.kelsos.mbrc.events.ui.ConnectionStatusChangeEvent;
import com.kelsos.mbrc.events.ui.CoverChangedEvent;
import com.kelsos.mbrc.events.ui.LfmRatingChanged;
import com.kelsos.mbrc.events.ui.OnMainFragmentOptionsInflated;
import com.kelsos.mbrc.events.ui.PlayStateChange;
import com.kelsos.mbrc.events.ui.RepeatChange;
import com.kelsos.mbrc.events.ui.ScrobbleChange;
import com.kelsos.mbrc.events.ui.ShuffleChange;
import com.kelsos.mbrc.events.ui.ShuffleChange.ShuffleState;
import com.kelsos.mbrc.events.ui.TrackInfoChangeEvent;
import com.kelsos.mbrc.events.ui.UpdatePosition;
import com.kelsos.mbrc.events.ui.VolumeChange;
import com.kelsos.mbrc.helper.VolumeChangeHelper;
import com.kelsos.mbrc.presenters.MainViewPresenter;
import com.kelsos.mbrc.ui.activities.BaseActivity;
import com.kelsos.mbrc.ui.dialogs.RatingDialogFragment;
import com.kelsos.mbrc.views.MainView;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;
import javax.inject.Inject;
import javax.inject.Singleton;
import timber.log.Timber;
import toothpick.Scope;
import toothpick.Toothpick;
import toothpick.smoothie.module.SmoothieActivityModule;

@Singleton
public class MainActivity extends BaseActivity implements MainView {
  private static final String PAUSED = "Paused";
  private static final String STOPPED = "Stopped";
  private final ScheduledExecutorService progressScheduler = Executors.newScheduledThreadPool(1);
  // Injects
  @Inject RxBus bus;
  @Inject MainViewPresenter presenter;
  // Inject elements of the view
  @BindView(R.id.main_artist_label) TextView artistLabel;
  @BindView(R.id.main_title_label) TextView titleLabel;
  @BindView(R.id.main_label_album) TextView albumLabel;
  @BindView(R.id.main_track_progress_current) TextView trackProgressCurrent;
  @BindView(R.id.main_track_duration_total) TextView trackDuration;
  @BindView(R.id.main_button_play_pause) ImageButton playPauseButton;
  @BindView(R.id.main_volume_seeker) SeekBar volumeBar;
  @BindView(R.id.main_track_progress_seeker) SeekBar progressBar;
  @BindView(R.id.main_mute_button) ImageButton muteButton;
  @BindView(R.id.main_shuffle_button) ImageButton shuffleButton;
  @BindView(R.id.main_repeat_button) ImageButton repeatButton;
  @BindView(R.id.main_album_cover_image_view) ImageView albumCover;
  private ShareActionProvider mShareActionProvider;
  private int previousVol;
  private ScheduledFuture mProgressUpdateHandler;
  private Menu menu;

  private SeekBar.OnSeekBarChangeListener progressBarChangeListener = new SeekBar.OnSeekBarChangeListener() {
    public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
      if (fromUser && progress != previousVol) {
        final UserAction action = new UserAction(Protocol.NowPlayingPosition, String.valueOf(progress));
        postAction(action);
        previousVol = progress;
      }
    }

    public void onStartTrackingTouch(SeekBar seekBar) {
    }

    public void onStopTrackingTouch(SeekBar seekBar) {
    }
  };
  private VolumeChangeHelper volumeChangeListener;
  private Scope scope;

  private void register() {
    this.bus.register(this, CoverChangedEvent.class, this::handleCoverEvent, true);
    this.bus.register(this, ShuffleChange.class, this::handleShuffleChange, true);
    this.bus.register(this, RepeatChange.class, this::updateRepeatButtonState, true);
    this.bus.register(this, VolumeChange.class, this::updateVolumeData, true);
    this.bus.register(this, PlayStateChange.class, this::handlePlayStateChange, true);
    this.bus.register(this, TrackInfoChangeEvent.class, this::handleTrackInfoChange, true);
    this.bus.register(this, ConnectionStatusChangeEvent.class, this::handleConnectionStatusChange, true);
    this.bus.register(this, UpdatePosition.class, this::handlePositionUpdate, true);
    this.bus.register(this, ScrobbleChange.class, this::handleScrobbleChange, true);
    this.bus.register(this, LfmRatingChanged.class, this::handleLfmLoveChange, true);
  }

  @OnClick(R.id.main_button_play_pause)
  void playButtonPressed() {
    final UserAction action = new UserAction(Protocol.PlayerPlayPause, true);
    postAction(action);
  }

  @OnClick(R.id.main_button_previous)
  void onPreviousButtonPressed() {
    final UserAction action = new UserAction(Protocol.PlayerPrevious, true);
    postAction(action);
  }

  @OnClick(R.id.main_button_next)
  void onNextButtonPressed() {
    final UserAction action = new UserAction(Protocol.PlayerNext, true);
    postAction(action);
  }

  @OnLongClick(R.id.main_button_play_pause)
  boolean onPlayerStopPressed() {
    final UserAction action = new UserAction(Protocol.PlayerStop, true);
    postAction(action);
    return true;
  }

  @OnClick(R.id.main_mute_button)
  void onMuteButtonPressed() {
    final UserAction action = new UserAction(Protocol.PlayerMute, Const.TOGGLE);
    postAction(action);
  }

  @OnClick(R.id.main_shuffle_button)
  void onShuffleButtonClicked() {
    final UserAction action = new UserAction(Protocol.PlayerShuffle, Const.TOGGLE);
    postAction(action);
  }

  @OnClick(R.id.main_repeat_button)
  void onRepeatButtonPressed() {
    final UserAction action = new UserAction(Protocol.PlayerRepeat, Const.TOGGLE);
    postAction(action);
  }

  /**
   * Posts a user action wrapped in a MessageEvent. The bus will
   * pass the MessageEvent through the Socket to the plugin.
   *
   * @param action Any kind of UserAction available in the {@link Protocol}
   */
  private void postAction(UserAction action) {
    bus.post(new MessageEvent(ProtocolEventType.UserAction, action));
  }

  private void changeVolume(int volume) {
    postAction(UserAction.create(Protocol.PlayerVolume, volume));
  }

  @Override
  public void onCreate(Bundle savedInstanceState) {
    scope = Toothpick.openScopes(getApplication(), this);
    scope.installModules(new SmoothieActivityModule(this));
    super.onCreate(savedInstanceState);
    Toothpick.inject(this, scope);
    setContentView(R.layout.activity_main);
    ButterKnife.bind(this);
    super.setup();
    volumeChangeListener = new VolumeChangeHelper(this::changeVolume);
    volumeBar.setOnSeekBarChangeListener(volumeChangeListener);
    progressBar.setOnSeekBarChangeListener(progressBarChangeListener);
  }

  @Override
  public void onStart() {
    super.onStart();
    setTextViewTypeface();
  }

  @Override
  public void onResume() {
    super.onResume();
    register();
    presenter.attach(this);
    presenter.requestNowPlayingPosition();
    presenter.load();
  }

  @Override
  protected void onPause() {
    super.onPause();
    presenter.detach();
  }

  @Override
  public boolean onOptionsItemSelected(MenuItem item) {
    switch (item.getItemId()) {
      case R.id.menu_lastfm_scrobble:
        presenter.toggleScrobbling();
        return true;
      case R.id.menu_rating_dialog:
        final RatingDialogFragment ratingDialog = new RatingDialogFragment();
        ratingDialog.show(getSupportFragmentManager(), "RatingDialog");
        return true;
      case R.id.menu_lastfm_love:
        bus.post(new MessageEvent(ProtocolEventType.UserAction,
            new UserAction(Protocol.NowPlayingLfmRating, Const.TOGGLE)));
        return true;
      default:
        return false;
    }
  }

  /**
   * Sets the typeface of the text views in the main activity to roboto.
   */
  private void setTextViewTypeface() {
    try {
      /* Marquee Hack */
      artistLabel.setSelected(true);
      titleLabel.setSelected(true);
      albumLabel.setSelected(true);

      Typeface robotoRegular = Typeface.createFromAsset(getAssets(), "fonts/roboto_regular.ttf");
      Typeface robotoMedium = Typeface.createFromAsset(getAssets(), "fonts/roboto_medium.ttf");
      Typeface robotoLight = Typeface.createFromAsset(getAssets(), "fonts/roboto_light.ttf");
      artistLabel.setTypeface(robotoRegular);
      titleLabel.setTypeface(robotoLight);
      albumLabel.setTypeface(robotoMedium);
      trackProgressCurrent.setTypeface(robotoMedium);
      trackDuration.setTypeface(robotoMedium);
    } catch (Exception ignore) {
      Timber.d(ignore, "Failed");
    }
  }

  @Override
  public void onStop() {
    super.onStop();
    bus.unregister(this);
  }

  @Override
  public boolean onCreateOptionsMenu(Menu menu) {
    getMenuInflater().inflate(R.menu.menu, menu);
    this.menu = menu;
    MenuItem shareItem = menu.findItem(R.id.actionbar_share);
    mShareActionProvider = (ShareActionProvider) MenuItemCompat.getActionProvider(shareItem);
    mShareActionProvider.setShareIntent(getShareIntent());
    bus.post(new OnMainFragmentOptionsInflated());
    return super.onCreateOptionsMenu(menu);
  }

  private Intent getShareIntent() {
    Intent shareIntent = new Intent(Intent.ACTION_SEND);
    shareIntent.setType("text/plain");
    final String payload = String.format("Now Playing: %s - %s", artistLabel.getText(), titleLabel.getText());
    shareIntent.putExtra(Intent.EXTRA_TEXT, payload);
    return shareIntent;
  }

  private void handleCoverEvent(final CoverChangedEvent cevent) {
    updateCover(cevent.getCover());
  }

  @Override
  public void updateCover(@Nullable Bitmap cover) {
    if (albumCover == null) {
      return;
    }
    if (cover != null) {
      albumCover.setImageBitmap(cover);
    } else {
      albumCover.setImageResource(R.drawable.ic_image_no_cover);
    }
  }

  private void handleShuffleChange(ShuffleChange change) {
    updateShuffleState(change.getShuffleState());
  }

  @Override
  public void updateShuffleState(@ShuffleState String shuffleState) {
    if (shuffleButton == null) {
      return;
    }

    final boolean shuffle = !ShuffleChange.OFF.equals(shuffleState);
    final boolean autoDj = ShuffleChange.AUTODJ.equals(shuffleState);

    int color = ContextCompat.getColor(this, shuffle ? R.color.accent : R.color.button_dark);
    shuffleButton.setColorFilter(color);

    shuffleButton.setImageResource(autoDj ? R.drawable.ic_headset_black_24dp : R.drawable.ic_shuffle_black_24dp);
  }

  private void updateRepeatButtonState(RepeatChange change) {
    updateRepeat(change.getMode());
  }

  @Override
  public void updateRepeat(@Mode String mode) {
    if (repeatButton == null) {
      return;
    }

    @ColorRes int colorId = R.color.accent;
    @DrawableRes int resId = R.drawable.ic_repeat_black_24dp;

    //noinspection StatementWithEmptyBody
    if (Repeat.ALL.equalsIgnoreCase(mode)) {
      // Do nothing already set above
    } else if (Repeat.ONE.equalsIgnoreCase(mode)) {
      resId = R.drawable.ic_repeat_one_black_24dp;
    } else {
      colorId = R.color.button_dark;
    }

    int color = ContextCompat.getColor(this, colorId);
    repeatButton.setImageResource(resId);
    repeatButton.setColorFilter(color);
  }

  private void updateVolumeData(VolumeChange change) {
    updateVolume(change.getVolume(), change.isMute());
  }

  @Override
  public void updateVolume(int volume, boolean mute) {
    if (volumeBar == null) {
      return;
    }

    if (!volumeChangeListener.isUserChangingVolume()) {
      volumeBar.setProgress(volume);
    }

    if (muteButton == null) {
      return;
    }

    int color = ContextCompat.getColor(this, R.color.button_dark);
    muteButton.setColorFilter(color);
    muteButton.setImageResource(mute ? R.drawable.ic_volume_off_black_24dp : R.drawable.ic_volume_up_black_24dp);
  }

  private void handlePlayStateChange(final PlayStateChange change) {
    updatePlayState(change.getState());
  }

  @Override
  public void updatePlayState(@State String state) {
    if (playPauseButton == null) {
      return;
    }
    int accentColor = ContextCompat.getColor(this, R.color.accent);
    @DrawableRes int resId;
    String tag;

    if (PlayerState.PLAYING.equals(state)) {
      resId = R.drawable.ic_pause_circle_filled_black_24dp;
      tag = "Playing";
        /* Start the animation if the track is playing*/
      presenter.requestNowPlayingPosition();
      trackProgressAnimation();
    } else if (PlayerState.PAUSED.equals(state)) {
      resId = R.drawable.ic_play_circle_filled_black_24dp;
      tag = PAUSED;
        /* Stop the animation if the track is paused*/
      stopTrackProgressAnimation();
    } else if (PlayerState.STOPPED.equals(state)) {
      resId = R.drawable.ic_play_circle_filled_black_24dp;
      tag = STOPPED;
        /* Stop the animation if the track is paused*/
      stopTrackProgressAnimation();
      activateStoppedState();
    } else {
      resId = R.drawable.ic_play_circle_filled_black_24dp;
      tag = STOPPED;
    }

    playPauseButton.setColorFilter(accentColor);
    playPauseButton.setImageResource(resId);
    playPauseButton.setTag(tag);
  }

  /**
   * If the track progress animation is running the the function stops it.
   */
  private void stopTrackProgressAnimation() {
    if (mProgressUpdateHandler != null) {
      mProgressUpdateHandler.cancel(true);
    }
  }

  /**
   * Starts the progress animation when called. If It was previously running then it restarts it.
   */
  private void trackProgressAnimation() {
    /* If the scheduled tasks is not null then cancel it and clear it along with the
    timer to create them anew */
    final int timePeriod = 1;
    stopTrackProgressAnimation();
    Object tag = playPauseButton.getTag();
    if (STOPPED.equals(tag) || PAUSED.equals(tag)) {
      return;
    }

    final Runnable updateProgress = () -> {

      int currentProgress = progressBar.getProgress() / 1000;
      final int currentMinutes = currentProgress / 60;
      final int currentSeconds = currentProgress % 60;

      runOnUiThread(() -> {
        try {
          if (progressBar == null) {
            return;
          }
          progressBar.setProgress(progressBar.getProgress() + 1000);
          trackProgressCurrent.setText(getString(R.string.playback_progress, currentMinutes, currentSeconds));
        } catch (Exception ex) {
          Timber.d(ex, "animation timer");
        }
      });
    };

    mProgressUpdateHandler = progressScheduler.scheduleAtFixedRate(updateProgress, 0, timePeriod, TimeUnit.SECONDS);
  }

  private void activateStoppedState() {
    if (trackProgressCurrent == null || progressBar == null) {
      return;
    }
    progressBar.setProgress(0);
    trackProgressCurrent.setText(getString(R.string.playback_progress, 0, 0));
  }

  private void handleTrackInfoChange(final TrackInfoChangeEvent change) {
    updateTrackInfo(change.getTrackInfo());
  }

  @Override
  public void updateTrackInfo(TrackInfo info) {
    if (artistLabel == null) {
      return;
    }
    artistLabel.setText(info.getArtist());
    titleLabel.setText(info.getTitle());
    albumLabel.setText(TextUtils.isEmpty(info.getYear())
        ? info.getAlbum()
        : String.format("%s [%s]", info.getAlbum(), info.getYear()));

    if (mShareActionProvider != null) {
      mShareActionProvider.setShareIntent(getShareIntent());
    }
  }

  private void handleConnectionStatusChange(final ConnectionStatusChangeEvent change) {
    updateConnection(change.getStatus());
  }

  @Override
  public void updateConnection(int status) {
    if (status == Connection.OFF) {
      stopTrackProgressAnimation();
      activateStoppedState();
    }
  }

  /**
   * Responsible for updating the displays and seekbar responsible for the display of the track
   * duration and the
   * current progress of playback
   */

  private void handlePositionUpdate(UpdatePosition position) {
    final int total = position.getTotal();
    final int current = position.getCurrent();
    if (trackProgressCurrent == null || progressBar == null || trackDuration == null) {
      return;
    }
    if (total == 0) {
      bus.post(new MessageEvent(UserInputEventType.RequestPosition));
      return;
    }
    int currentSeconds = current / 1000;
    int totalSeconds = total / 1000;

    final int currentMinutes = currentSeconds / 60;
    final int totalMinutes = totalSeconds / 60;

    currentSeconds %= 60;
    totalSeconds %= 60;
    final int finalTotalSeconds = totalSeconds;
    final int finalCurrentSeconds = currentSeconds;

    trackDuration.setText(getString(R.string.playback_progress, totalMinutes, finalTotalSeconds));
    trackProgressCurrent.setText(getString(R.string.playback_progress, currentMinutes, finalCurrentSeconds));

    progressBar.setMax(total);
    progressBar.setProgress(current);

    trackProgressAnimation();
  }

  private void handleScrobbleChange(ScrobbleChange event) {
    updateScrobbleStatus(event.isActive());
  }

  @Override
  public void updateScrobbleStatus(boolean active) {
    if (menu == null) {
      return;
    }
    final MenuItem scrobbleMenuItem = menu.findItem(R.id.menu_lastfm_scrobble);
    if (scrobbleMenuItem == null) {
      return;
    }

    scrobbleMenuItem.setChecked(active);
  }

  private void handleLfmLoveChange(LfmRatingChanged event) {
    updateLfmStatus(event.getStatus());
  }

  @Override
  public void updateLfmStatus(LfmStatus status) {
    if (menu == null) {
      return;
    }
    final MenuItem favoriteMenuItem = menu.findItem(R.id.menu_lastfm_love);
    if (favoriteMenuItem == null) {
      return;
    }

    switch (status) {
      case LOVED:
        favoriteMenuItem.setIcon(R.drawable.ic_favorite_black_24dp);
        break;
      default:
        favoriteMenuItem.setIcon(R.drawable.ic_favorite_border_black_24dp);
        break;
    }
  }

  @Override
  protected int active() {
    return R.id.nav_home;
  }

  @Override
  protected void onDestroy() {
    Toothpick.closeScope(this);
    super.onDestroy();
  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/ui/navigation/main/MainActivity.kt
package com.kelsos.mbrc.ui.navigation.main

import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.os.Bundle
import androidx.annotation.ColorRes
import androidx.annotation.DrawableRes
import androidx.core.content.ContextCompat
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.widget.ShareActionProvider
import android.text.TextUtils
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.SeekBar
import android.widget.TextView
import com.kelsos.mbrc.R
import com.kelsos.mbrc.content.activestatus.PlayerState
import com.kelsos.mbrc.content.activestatus.PlayerState.State
import com.kelsos.mbrc.content.activestatus.PlayerStatusModel
import com.kelsos.mbrc.content.activestatus.Repeat
import com.kelsos.mbrc.content.activestatus.Repeat.Mode
import com.kelsos.mbrc.content.activestatus.TrackPositionData
import com.kelsos.mbrc.content.activestatus.TrackRatingModel
import com.kelsos.mbrc.content.library.tracks.PlayingTrackModel
import com.kelsos.mbrc.events.ShuffleMode
import com.kelsos.mbrc.events.ShuffleMode.Shuffle
import com.kelsos.mbrc.extensions.getDimens
import com.kelsos.mbrc.networking.connections.Connection
import com.kelsos.mbrc.ui.activities.BaseNavigationActivity
import com.kelsos.mbrc.ui.dialogs.RatingDialogFragment
import com.kelsos.mbrc.ui.navigation.main.ProgressSeekerHelper.ProgressUpdate
import com.squareup.picasso.Picasso
import kotterknife.bindView
import toothpick.Scope
import toothpick.Toothpick
import toothpick.smoothie.module.SmoothieActivityModule
import java.io.File
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class MainActivity : BaseNavigationActivity(), MainView, ProgressUpdate {

  // Injects
  @Inject
  lateinit var presenter: MainViewPresenter
  @Inject
  lateinit var progressHelper: ProgressSeekerHelper
  // Inject elements of the view
  private val artistLabel: TextView by bindView(R.id.main_artist_label)
  private val titleLabel: TextView by bindView(R.id.main_title_label)
  private val albumLabel: TextView by bindView(R.id.main_label_album)
  private val trackProgressCurrent: TextView by bindView(R.id.main_track_progress_current)
  private val trackDuration: TextView by bindView(R.id.main_track_duration_total)
  private val playPauseButton: ImageButton by bindView(R.id.main_button_play_pause)
  private val volumeBar: SeekBar by bindView(R.id.main_volume_seeker)
  private val progressBar: SeekBar by bindView(R.id.main_track_progress_seeker)
  private val muteButton: ImageButton by bindView(R.id.main_mute_button)
  private val shuffleButton: ImageButton by bindView(R.id.main_shuffle_button)
  private val repeatButton: ImageButton by bindView(R.id.main_repeat_button)
  private val albumCover: ImageView by bindView(R.id.main_album_cover_image_view)
  private val previousButton: ImageButton by bindView(R.id.main_button_previous)
  private val nextButton: ImageButton by bindView(R.id.main_button_next)
  private val trackInfo: View by bindView(R.id.track_info_area)

  private var mShareActionProvider: ShareActionProvider? = null

  private var changeLogDialog: AlertDialog? = null
  private var outOfDateDialog: AlertDialog? = null

  private var menu: Menu? = null
  private var volumeChangeListener: SeekBarThrottler? = null

  private var positionChangeListener: SeekBarThrottler? = null
  private lateinit var scope: Scope

  override fun onCreate(savedInstanceState: Bundle?) {
    scope = Toothpick.openScopes(application, PRESENTER_SCOPE, this)
    scope.installModules(SmoothieActivityModule(this), MainModule())
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_main)
    Toothpick.inject(this, scope)
    playPauseButton.setOnClickListener { presenter.play() }
    playPauseButton.setOnLongClickListener { presenter.stop() }
    previousButton.setOnClickListener { presenter.previous() }
    nextButton.setOnClickListener { presenter.next() }
    muteButton.setOnClickListener { presenter.mute() }
    shuffleButton.setOnClickListener { presenter.shuffle() }
    repeatButton.setOnClickListener { presenter.repeat() }
    trackInfo.setOnClickListener { navigate(R.id.nav_now_playing) }

    super.setup()
    presenter.attach(this)
    presenter.load()
  }

  override fun onNewIntent(intent: Intent?) {
    super.onNewIntent(intent)
    if (intent?.getBooleanExtra(EXIT_APP, false) == true) {
      exitApplication()
      return
    }
  }

  override fun showChangeLog() {
    changeLogDialog = AlertDialog.Builder(this)
      .setTitle(R.string.main__dialog_change_log)
      .setView(R.layout.change_log_dialog)
      .setPositiveButton(android.R.string.ok) { dialogInterface, _ -> dialogInterface.dismiss() }
      .show()
  }

  override fun notifyPluginOutOfDate() {
    outOfDateDialog = AlertDialog.Builder(this)
      .setTitle(R.string.main__dialog_plugin_outdated_title)
      .setMessage(R.string.main__dialog_plugin_outdated_message)
      .setPositiveButton(android.R.string.ok) { dialogInterface, _ -> dialogInterface.dismiss() }
      .show()
  }

  override fun onStart() {
    super.onStart()

    if (!presenter.isAttached) {
      presenter.attach(this)
      presenter.load()
    }

    progressHelper.setProgressListener(this)
    volumeChangeListener = SeekBarThrottler { presenter.changeVolume(it) }
    positionChangeListener = SeekBarThrottler { presenter.seek(it) }
    volumeBar.setOnSeekBarChangeListener(volumeChangeListener)
    progressBar.setOnSeekBarChangeListener(positionChangeListener)
    artistLabel.isSelected = true
    titleLabel.isSelected = true
    albumLabel.isSelected = true
  }

  override fun onResume() {
    super.onResume()
    presenter.requestNowPlayingPosition()
  }

  override fun onOptionsItemSelected(item: MenuItem): Boolean {
    return when (item.itemId) {
      R.id.menu_lastfm_scrobble -> {
        presenter.toggleScrobbling()
        true
      }
      R.id.menu_rating_dialog -> {
        RatingDialogFragment.create(this).show()
        true
      }
      R.id.menu_lastfm_love -> {
        presenter.lfmLove()
      }
      else -> false
    }
  }

  override fun onStop() {
    super.onStop()
    presenter.detach()
    progressHelper.setProgressListener(null)
    volumeChangeListener?.terminate()
    volumeChangeListener = null
    positionChangeListener?.terminate()
    positionChangeListener = null
    volumeBar.setOnSeekBarChangeListener(null)
    progressBar.setOnSeekBarChangeListener(null)
  }

  override fun onCreateOptionsMenu(menu: Menu): Boolean {
    menuInflater.inflate(R.menu.menu, menu)
    this.menu = menu
//    val shareItem = menu.findItem(R.id.actionbar_share)
//    mShareActionProvider = MenuItemCompat.getActionProvider(shareItem) as ShareActionProvider
//    mShareActionProvider!!.setShareIntent(shareIntent)
    //bus.post(OnMainFragmentOptionsInflated())
    return super.onCreateOptionsMenu(menu)
  }

  private val shareIntent: Intent
    get() {
      val shareIntent = Intent(Intent.ACTION_SEND)
      shareIntent.type = "text/plain"
      val payload = String.format("Now Playing: %s - %s", artistLabel.text, titleLabel.text)
      shareIntent.putExtra(Intent.EXTRA_TEXT, payload)
      return shareIntent
    }

  override fun updateRating(rating: TrackRatingModel) {
    menu?.run {
      val favoriteMenuItem = findItem(R.id.menu_lastfm_love) ?: return

      when (rating.lfmRating) {
        LfmRating.LOVED -> favoriteMenuItem.setIcon(R.drawable.ic_favorite_black_24dp)
        else -> favoriteMenuItem.setIcon(R.drawable.ic_favorite_border_black_24dp)
      }
    }
  }

  override fun updateStatus(playerStatus: PlayerStatusModel) {
    with(playerStatus) {
      updateShuffleState(shuffle)
      updateRepeat(repeat)
      updateVolume(volume, mute)
      updatePlayState(playState)
      updateScrobbleStatus(scrobbling)
    }
  }

  private fun updateCover(path: String) {
    val file = File(path)

    if (!file.exists()) {
      albumCover.setImageResource(R.drawable.ic_image_no_cover)
      return
    }

    if (albumCover.tag == path) {
      return
    }

    albumCover.tag = path

    val dimens = getDimens()

    Picasso.get()
      .load(file)
      .noFade()
      .error(R.drawable.ic_image_no_cover)
      .config(Bitmap.Config.RGB_565)
      .resize(dimens, dimens)
      .centerCrop()
      .into(albumCover)
  }

  private fun updateShuffleState(@Shuffle shuffleState: String) {
    val shuffle = ShuffleMode.OFF != shuffleState
    val autoDj = ShuffleMode.AUTODJ == shuffleState

    val colorResId = if (shuffle) R.color.accent else R.color.button_dark
    val iconResId =
      if (autoDj) R.drawable.ic_headset_black_24dp else R.drawable.ic_shuffle_black_24dp

    val color = ContextCompat.getColor(this, colorResId)
    shuffleButton.setColorFilter(color)
    shuffleButton.setImageResource(iconResId)
  }

  private fun updateRepeat(@Mode mode: String) {
    @ColorRes var colorId = R.color.accent
    @DrawableRes var resId = R.drawable.ic_repeat_black_24dp

    //noinspection StatementWithEmptyBody
    when {
      Repeat.ALL.equals(mode, ignoreCase = true) -> Unit
      Repeat.ONE.equals(mode, ignoreCase = true) -> resId = R.drawable.ic_repeat_one_black_24dp
      else -> colorId = R.color.button_dark
    }

    val color = ContextCompat.getColor(this, colorId)
    repeatButton.setImageResource(resId)
    repeatButton.setColorFilter(color)
  }

  private fun updateVolume(volume: Int, mute: Boolean) {
    val color = ContextCompat.getColor(this, R.color.button_dark)
    val iconResId = if (mute) {
      R.drawable.ic_volume_off_black_24dp
    } else {
      R.drawable.ic_volume_up_black_24dp
    }
    volumeBar.progress = volume
    with(muteButton) {
      setColorFilter(color)
      setImageResource(iconResId)
    }
  }

  private fun updatePlayState(@State state: String) {
    val accentColor = ContextCompat.getColor(this, R.color.accent)
    @DrawableRes val resId: Int
    val tag: String

    when (state) {
      PlayerState.PLAYING -> {
        resId = R.drawable.ic_pause_circle_filled_black_24dp
        tag = "Playing"
        /* Start the animation if the track is playing*/
        presenter.requestNowPlayingPosition()
        trackProgressAnimation(progressBar.progress, progressBar.max)
      }
      PlayerState.PAUSED -> {
        resId = R.drawable.ic_play_circle_filled_black_24dp
        tag = PAUSED
        /* Stop the animation if the track is paused*/
        progressHelper.stop()
      }
      PlayerState.STOPPED -> {
        resId = R.drawable.ic_play_circle_filled_black_24dp
        tag = STOPPED
        /* Stop the animation if the track is paused*/
        progressHelper.stop()
        activateStoppedState()
      }
      else -> {
        resId = R.drawable.ic_play_circle_filled_black_24dp
        tag = STOPPED
      }
    }

    playPauseButton.setColorFilter(accentColor)
    playPauseButton.setImageResource(resId)
    playPauseButton.tag = tag
  }

  /**
   * Starts the progress animation when called. If It was previously running then it restarts it.
   */
  private fun trackProgressAnimation(current: Int, total: Int) {
    progressHelper.stop()

    val tag = playPauseButton.tag
    if (STOPPED == tag || PAUSED == tag) {
      return
    }

    progressHelper.update(current, total)
  }

  private fun activateStoppedState() {
    progressBar.progress = 0
    trackProgressCurrent.text = getString(R.string.playback_progress, 0, 0)
  }

  override fun updateTrackInfo(info: PlayingTrackModel) {
    artistLabel.text = info.artist
    titleLabel.text = info.title
    albumLabel.text = if (TextUtils.isEmpty(info.year)) info.album
    else String.format("%s [%s]", info.album, info.year)

    if (mShareActionProvider != null) {
      mShareActionProvider!!.setShareIntent(shareIntent)
    }

    updateCover(info.coverUrl)
  }

  override fun updateConnection(status: Int) {
    if (status == Connection.OFF) {
      progressHelper.stop()
      activateStoppedState()
    }
  }

  /**
   * Responsible for updating the displays and seekbar responsible for the display of the track
   * duration and the
   * current progress of playback
   */

  override fun updateProgress(position: TrackPositionData) {
    val total = position.total
    val current = position.current

    var currentSeconds = current / 1000
    var totalSeconds = total / 1000

    val currentMinutes = currentSeconds / 60
    val totalMinutes = totalSeconds / 60

    currentSeconds %= 60
    totalSeconds %= 60
    val finalTotalSeconds = totalSeconds
    val finalCurrentSeconds = currentSeconds

    trackDuration.text = getString(R.string.playback_progress, totalMinutes, finalTotalSeconds)
    trackProgressCurrent.text = getString(
      R.string.playback_progress,
      currentMinutes,
      finalCurrentSeconds
    )

    progressBar.max = total
    progressBar.progress = current

    trackProgressAnimation(position.current, position.total)
  }

  private fun updateScrobbleStatus(active: Boolean) {
    if (menu == null) {
      return
    }
    val scrobbleMenuItem = menu!!.findItem(R.id.menu_lastfm_scrobble) ?: return

    scrobbleMenuItem.isChecked = active
  }

  override fun active(): Int {
    return R.id.nav_home
  }

  override fun progress(position: Int, duration: Int) {
    val currentProgress = progressBar.progress / 1000
    val currentMinutes = currentProgress / 60
    val currentSeconds = currentProgress % 60

    progressBar.progress = progressBar.progress + 1000
    trackProgressCurrent.text = getString(
      R.string.playback_progress,
      currentMinutes,
      currentSeconds
    )
  }

  override fun onDestroy() {
    Toothpick.closeScope(this)
    if (isFinishing) {
      //when we leave the presenter flow,
      //we close its scope
      Toothpick.closeScope(PRESENTER_SCOPE)
    }
    outOfDateDialog?.dismiss()
    changeLogDialog?.dismiss()
    super.onDestroy()
  }

  @javax.inject.Scope
  @Target(AnnotationTarget.TYPE)
  @Retention(AnnotationRetention.RUNTIME)
  annotation class Presenter

  companion object {
    fun start(context: Context) {
      with(context) {
        startActivity(Intent(this, MainActivity::class.java))
      }
    }

    private const val PAUSED = "Paused"
    private const val STOPPED = "Stopped"
    private val PRESENTER_SCOPE: Class<*> = Presenter::class.java
  }
}